src/gui/widgets/spinbox.cpp

```diff
@@ -0,0 +1,180 @@
+/*
+	Copyright (C) 2008 - 2024
+	by babaissarkar(Subhraman Sarkar) <suvrax@gmail.com>
+	Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+	This program is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY.
+
+	See the COPYING file for more details.
+*/
+
+#define GETTEXT_DOMAIN "wesnoth-lib"
+
+#include "gui/widgets/spinbox.hpp"
+
+#include "gui/auxiliary/find_widget.hpp"
+#include "gui/widgets/repeating_button.hpp"
+#include "gui/widgets/text_box.hpp"
+#include "gui/core/log.hpp"
+#include "gui/core/window_builder/helper.hpp"
+#include "gui/core/register_widget.hpp"
+#include "gui/widgets/settings.hpp"
+#include "gui/widgets/window.hpp"
+#include "gettext.hpp"
+#include "wml_exception.hpp"
+
+#include <functional>
+
+#define LOG_SCOPE_HEADER get_control_type() + " [" + id() + "] " + __func__
+#define LOG_HEADER LOG_SCOPE_HEADER + ':'
+
+namespace gui2
+{
+
+// ------------ WIDGET -----------{
+
+REGISTER_WIDGET(spinner)
+
+spinner::spinner(const implementation::builder_spinner& builder)
+	: container_base(builder, type())
+	, state_(ENABLED)
+	, step_size_(1)
+	, invalid_(false)
+{
+	connect_signal<event::LEFT_BUTTON_DOWN>(
+		std::bind(&spinner::signal_handler_left_button_down, this, std::placeholders::_2),
+		event::dispatcher::back_pre_child);
+}
+
+text_box* spinner::get_internal_text_box()
+{
+	return find_widget<text_box>(this, "_text", false, true);
+}
+
+void spinner::set_value(const int val)
+{
+	text_box* edit_area = get_internal_text_box();
+	if (edit_area != nullptr) {
+		edit_area->set_value(std::to_string(val));
+	}
+}
+
+int spinner::get_value()
+{
+	/* Return 0 if invalid.
+	 * TODO: give visual indication of wrong value
```

Comment:
I'm happy to leave this as a todo, but just checking that it's deliberately left like that.

Reply:
Yes, might be added in the future.